src/ast.rs

```diff
@@ -491,9 +494,49 @@ fn resolve(db: &dyn Ast, location: NodeLocation) -> Option<Arc<Decl>> {
                 _ => return None,
             }
         }
+        "for_expr" => {
+            let children = node
+                .named_children_not("nl")
+                .into_iter()
+                .collect::<Vec<_>>();
+            let loop_vars = &children[0];
+            let loop_expr = &children[1];
+
+            let expr_type = db.resolve(NodeLocation::from_node(uri.clone(), *loop_expr))?;
+            let expr_type = db.typ(expr_type)?;
+
+            match &expr_type.kind {
+                DeclKind::Builtin(Type::Set(element_types)) => {
+                    let mut resolved_vars = Vec::new();
+                    for (i, typ) in element_types.iter().enumerate() {
+                        // Loop variables should be resolved based on order
+                        let var_name = loop_vars
+                            .named_child(&i.to_string())
+                            .and_then(|n| {
+                                n.utf8_text(source.as_bytes()).ok().map(ToString::to_string)
+                            })
+                            .unwrap_or_else(|| format!("var_{i}_{typ:?}"));
+                        let var_decl = Arc::new(Decl {
+                            module: query::ModuleId::Global,
+                            id: var_name.clone().into(),
+                            fqid: var_name.into(),
+                            kind: DeclKind::LoopIndex(i, expr_type.id.to_string()),
+                            is_export: None,
+                            loc: Some(Location::from(NodeLocation::from_node(
```

Comment:
`Into` and `From` work together, so I'd personally go with `NodeLocation::from_node(...).into()` instead of using `Location::from`. This would also allow you to drop the import of `Location` above. No need to change this as I believe this code would need to go away.